akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickProjectionImpl.scala

```diff
@@ -179,38 +183,80 @@ private[projection] class SlickProjectionImpl[Offset, Envelope, P <: JdbcProfile
       def processEnvelopeAndStoreOffsetInSameTransaction(
           handler: SlickHandler[Envelope],
           handlerRecovery: HandlerRecoveryImpl[Offset, Envelope],
+          offset: Offset,
           env: Envelope): Future[Done] = {
-        val offset = sourceProvider.extractOffset(env)
-        // run user function and offset storage on the same transaction
-        // any side-effect in user function is at-least-once
-        val txDBIO =
-          offsetStore
-            .saveOffset(projectionId, offset)
-            .flatMap(_ => handler.process(env))
-            .transactionally
-
-        handlerRecovery.applyRecovery(env, offset, offset, () => databaseConfig.db.run(txDBIO).map(_ => Done))
+        handlerRecovery.applyRecovery(env, offset, offset, {
+          () =>
+            val handlerAction = handler.process(env)
```

Comment:
> Now the order of the calls `process` and `saveOffset` was changed. Shouldn't matter as long as `process` is free from external side effects, but otherwise it's better to first saveOffset and then process.
> 
> Can't we move the verification to right before `databaseConfig.db.run(txDBIO)`?

This won't work because `saveOffset` is deferred execution.  It isn't actually run until it's passed to `databaseConfig.db.run`.  That means that running the envelope handler only ran during db `run` too and there's no opportunity to verify the offset after the envelope handler is executed.

We could flatmap in the `verifyOffset` after running the envelope handler, but then the challenge is parsing the resulting `DBIOAction.failed` to check if it was offset verification failure and skip records, or some other kind of failure and let the recovery strategy takeover.

Since the `saveOffset` and `DBIOAction` returned by the envelope handler are processed transactionally the order doesn't really matter, but if the user produces any other side effects in the envelope handler at least we'll know those are executed before we verify the offset the second time.

Reply:
I see. It's fine to change the order of the calls.

Reply:
Sorry, I'm missing some context here. Why do we need to process the handler before the verification? 

My understanding was, when the time comes to process the envelope (user + saveOffset) we check with the SourceProvider to know if that in-flight envelope still need to be processed. If so, go ahead, if not skip it. 

I don't see why we need to build the user DBIO before it. 

And because we are talking about potential side-effect in user code, this will make the side-effect to run twice. Once in the projection that should have skipped it and then again on the projection that will receive that envelope once the hand-over happens.

Reply:
There is already a verify earlier in the source before calling the handler. The idea here was that it should also be verified after calling the handler. I think it could be better to actually verify as part of the `run`, and rollback if it fails. Then we do it as late as possible.

```
val txDBIO =
  offsetStore
    .saveOffset(projectionId, offset)
    .flatMap(_ => handler.process(env))
    .flatMap { _ =>
      sourceProvider.verifyOffset(offset) match {
        case VerificationSuccess => DBIO.success...
        case VerificationFailure(reason) => DBIO.failure...
    }
    .transactionally
```

then after `run` recover the VerificationFailure exception and return Future.successful to ignore.

wdyt @seglo ?